### PR TITLE
Model factory of blocks

### PR DESCRIPTION
**Pull request by lucidtronix**
Line 1749 is where the high-level model factory is defined.

### CODE REVIEW

File: ml4h/models.py [hunk 1]
```diff
@@ -842,13 +868,409 @@ def __call__(self, x: Tensor, intermediates: Dict[TensorMap, List[Tensor]], _) -
...
+            self,
+            tensor_map: TensorMap,
+    ):
+        self.tensor_map = tensor_map
+        if not self.can_apply():
+            return
+        self.dense = Dense(tensor_map.shape[-1], activation=tensor_map.activation, name=tensor_map.output_name())
+
+    def can_apply(self):
+        return self.tensor_map.is_language()
```
Comment (ndiamant) [inline ml4h/models.py:957]: why is this necessary? Can't a fully connected layer be applied to anything?

File: ml4h/models.py [hunk 2]
```diff
@@ -842,13 +868,409 @@ def __call__(self, x: Tensor, intermediates: Dict[TensorMap, List[Tensor]], _) -
...
+            widths=dense_layers,
+            activation=activation,
+            normalization=dense_normalize,
+            regularization=dense_regularize,
+            regularization_rate=dense_regularize_rate,
+            name=f'embed_{self.tensor_map.name}',
+        ) if dense_layers else None
+
+    def can_apply(self):
+        return self.tensor_map.axes() > 1
```
Comment (ndiamant) [inline ml4h/models.py:1028]: Could this also check that the axes < 4 (since no 4d convolution)?

File: ml4h/models.py [hunk 3]
```diff
@@ -842,13 +868,409 @@ def __call__(self, x: Tensor, intermediates: Dict[TensorMap, List[Tensor]], _) -
...
+            x = self.fully_connected(x)
+            intermediates[self.tensor_map].append(x)
+        return x
+
+
+class ConvDecoderBlock:
+    def __init__(
+            self,
+            *,
+            tensor_map: TensorMap,
```
Comment (ndiamant) [inline ml4h/models.py:1051]: If I understand right, this is the output tensormap for a model, so maybe call it `tensor_map_out`

File: ml4h/models.py [hunk 4]
```diff
@@ -1145,31 +1563,62 @@ def _make_multimodal_multitask_model(
     return Model(inputs=list(inputs.values()), outputs=list(decoder_outputs.values()))
 
 
+def _transfer_layers_by_name(model_layers: str, freeze_model_layers: str, custom_dict: Dict[str, Any], m: Model):
```
Comment (ndiamant) [inline ml4h/models.py:1566]: thanks for putting this in its own function!

File: ml4h/models.py [hunk 5]
```diff
@@ -1273,6 +1722,157 @@ def make_paired_autoencoder_model(
...
+    return full_model, encoders, decoders, merger
+
+
+def _make_multimodal_multitask_model_block(
+        encoder_block_functions: Dict[TensorMap, Block],
+        merge: Block,
+        decoder_block_functions: Dict[TensorMap, Block],  # Assumed to be topologically sorted according to parents hierarchy
+        u_connect: DefaultDict[TensorMap, Set[TensorMap]],
+) -> Tuple[Model, Dict[TensorMap, Model], Dict[TensorMap, Model]]:
+    inputs: Dict[TensorMap, Input] = {}
```
Comment (ndiamant) [inline ml4h/models.py:1843]: docstring would be helpful here

File: ml4h/models.py [hunk 6]
```diff
@@ -1273,6 +1722,157 @@ def make_paired_autoencoder_model(
...
+    )
+
+    if kwargs.get('model_file', False):
+        return _load_model_encoders_and_decoders(tensor_maps_in, tensor_maps_out, custom_dict, opt, kwargs['model_file'])
+
+    encoder_block_functions = {tm: identity for tm in tensor_maps_in}  # Dict[TensorMap, Block]
+    for tm in tensor_maps_in:
+        for encode_block in encoder_blocks:
+            if encode_block in BLOCK_CLASSES:
+                encoder_block_functions[tm] = compose(encoder_block_functions[tm], BLOCK_CLASSES[encode_block](tensor_map=tm, **kwargs))
```
Comment (ndiamant) [inline ml4h/models.py:1799]: Does each `encode_block` get  applied to each input it `can_apply` to?
Is it possible to encode the scenario where I want to apply 2 dense blocks to an ECG and 1 dense block to an MRI?

File: ml4h/models.py [hunk 7]
```diff
@@ -1273,6 +1722,157 @@ def make_paired_autoencoder_model(
...
+            elif encode_block.endswith(f'encoder_{tm.name}.h5'):
```
Comment (ndiamant) [inline ml4h/models.py:1800]: Make this naming requirement explicit elsewhere

File: ml4h/models.py [hunk 8]
```diff
@@ -1273,6 +1722,157 @@ def make_paired_autoencoder_model(
...
+    The tasks attempted are given by the output TensorMaps.
+    The modalities and the first layers in the architecture are determined by the input TensorMaps.
+    Model architectures are specified by Blocks which can encode, merge or decode TensorMaps.
+
+    Hyperparameters are exposed to the command line and passed through to Block constructors via **kwargs.
+    Model summary printed to output
+
+    :param tensor_maps_in: List of input TensorMaps
+    :param tensor_maps_out: List of output TensorMaps
+    :param encoder_blocks: One or more BLOCK_CLASS keys or hd5 model files to encode applicable input TensorMaps
```
Comment (ndiamant) [inline ml4h/models.py:1774]: Could this instead be a dictionary from tensormap to list of blocks to apply to that tensormap?
E.g. 
```
{
  ecg_tmap: [denseblock(32 channels), denseblock(64 channels)]
  age_tmap: [fullyconnected_block(32 units)]
}
```

**Review by ndiamant — CHANGES_REQUESTED** [review]
I'm reviewing only the model block code, not the tf2.3 or paired autoencoder stuff.
Since this is an early PR, just focusing on the API design and style.

A lot of the blocks follow the structure of having a `__call__` and a `can_apply` function. Could you make an abstract base class with comments?

**Comment by ndiamant** [comment]
Recording how `TensorMap` is used
```
tm.shape, name=tm.input_name()
losses.append(tm.loss)
loss_weights.append(tm.loss_weight)
tm.output_name()
            if tm.axes() > 1:
        if not tm.parents or set(tm.parents) <= set(final):
if not tm.is_embedding()
        if tm.is_language():
        metrics={tm.output_name(): tm.metrics for tm in tensor_maps_out},
activation=tensor_map.activation
        self.units = tensor_map.annotation_units
```

File: ml4h/models.py (hunk 6 above)
Comment (lucidtronix) [reply ml4h/models.py:1799]: Yes,  but for now they would have to be serialized separately

File: ml4h/models.py (hunk 4 above)
Comment (lucidtronix) [reply ml4h/models.py:1566]: :)

File: ml4h/models.py (hunk 3 above)
Comment (lucidtronix) [reply ml4h/models.py:1051]: we could, but it's kinda nice to have just tensor_map for Blocks that can be encoders or decoders not to have to specify `_in` or `_out`

**Comment by ndiamant** [comment]
01/28/2021 meeting conclusions

- [ ]  don't compile model in model factory
- [ ] Move model factory to model directory, move blocks to separate files in model directory
- [ ] Allow model factory to take python objects for blocks, not just strings
- [ ] Model factory wrapper handles the command line, combining optimizers, loading model files, and finding the right blocks

File: ml4h/models/model_factory.py [hunk 9]
```diff
@@ -0,0 +1,282 @@
...
+            elif decode_block.endswith(f'decoder_{tm.name}.h5'):
+                serialized_decoder = load_model(decode_block, custom_objects=custom_dict, compile=False)
+                decoder_block_functions[tm] = compose(decoder_block_functions[tm], ModelAsBlock(tensor_map=tm, model=serialized_decoder))
+                break
+
+    return _make_multimodal_multitask_model_block(encoder_block_functions, merge, decoder_block_functions, u_connect)
+
+
+def _make_multimodal_multitask_model_block(
+        encoder_block_functions: Dict[TensorMap, Block],
```
Comment (ndiamant) [inline ml4h/models/model_factory.py:182]: If we want one block per tensormap, can we have a `BlockFromBlocks` class that chains blocks together?
That way you could have an encoder be a series of 3 residual blocks

**Review by ndiamant — COMMENTED** [review]
This is looking great! I think there are is still some commented out code from another branch, but once that's removed, I think it's ready.

File: ml4h/models/model_factory.py (hunk 9 above)
Comment (lucidtronix) [reply ml4h/models/model_factory.py:182]: Yes we could but it also works currently with lists of Blocks via the 
```
def compose(f, g):
    return lambda x, y: g(f(x, y), y)
```
then during model building calling: `compose(encoder_block_functions[tm], encode_block(tensor_map=tm, **kwargs))`

File: ml4h/models/model_factory.py [hunk 10]
```diff
@@ -0,0 +1,282 @@
...
+                ))
+            elif decode_block.endswith(f'decoder_{tm.name}.h5'):
+                serialized_decoder = load_model(decode_block, custom_objects=custom_dict, compile=False)
+                decoder_block_functions[tm] = compose(decoder_block_functions[tm], ModelAsBlock(tensor_map=tm, model=serialized_decoder))
+                break
+
+    return _make_multimodal_multitask_model_block(encoder_block_functions, merge, decoder_block_functions, u_connect)
+
+
+def _make_multimodal_multitask_model_block(
```
Comment (mklarqvist) [inline ml4h/models/model_factory.py:181]: Make this public and we good to go

**Review by mklarqvist — APPROVED** [review]
